File: October 2022/Day 28/C++ Code/49. Group Anagrams.cpp
class Solution {
public:
    
    // Time Complexity :- O(n.m.log(n));
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        int n = strs.size();
        vector<pair<string,int>> s1;
        for(int i=0;i<n;i++) {
            pair<string,int> p(sortString(strs[i]),i);
            s1.push_back(p);
        }
        
        sort(s1.begin(),s1.end(),comp);
        
        vector<vector<string>> ans;
        vector<string> list;
        for(int i=0;i<n;i++) {
            list.push_back(strs[s1[i].second]);
            if(i!=n-1 && s1[i].first.compare(s1[i+1].first)!=0) {
                ans.push_back(list);
                list.clear();
            }
        }
        ans.push_back(list);
        return ans;
    }
    
    // Time Compleixty :- O(m)
    static bool comp(const pair<string,int>& p1,const pair<string,int>& p2) {
        return p1.first<p2.first;
    }
    
    // Time Complexity :- O(m.log(m))
    string sortString(string str) {
        sort(str.begin(),str.end());
        return str;
    }
    
};

File: October 2022/Day 28/JAVA Code/49. Group Anagrams.java
class Solution {
    
    // Time Complexity :- O(n.m.log(n));
    
    public List<List<String>> groupAnagrams(String[] strs) {
        int n = strs.length;
        String[][] s1 = new String[n][2];
        for(int i=0;i<n;i++) {
            s1[i][0]= sort(strs[i]);
            s1[i][1]=Integer.toString(i);
        }
        
        Arrays.sort(s1,(a,b) -> compare(a[0],b[0]));
        
        List<List<String>> ans = new ArrayList<>();
        List<String> list = new ArrayList<>();
        for(int i=0;i<n;i++) {
            int idx = Integer.parseInt(s1[i][1]);
            list.add(strs[idx]);
            if(i!=n-1 && compare(s1[i][0],s1[i+1][0])!=0) {
                ans.add(list);
                list = new ArrayList<>();
            }
        }
        ans.add(list);
        return ans;
    }
    
    // Time Compleixty :- O(m)
    public static int compare(String s1,String s2) {
        return s1.compareTo(s2);
    }
    
    // Time Complexity :- O(m.log(m))
    public static String sort(String s) {
        char arr[] = s.toCharArray();
        Arrays.sort(arr);
        return new String(arr);
    }
}
